doc/specs/#885 - Terminal Settings Model/Actions Addendum.md

```diff
@@ -0,0 +1,334 @@
+    - `ActionMap` must be responsible for changing `Action`s:
+        - If `Name` changes, update `_NameMap`
+        - If `Keys` changes, update `_KeyMap`
+        - Also update the `Action` itself
+    - This is similar to how color schemes are maintained today.
+    - In the event that name/key-chord is set to something that's already taken, we need to propagate those changes to
+       the rest of `ActionMap`. As we do with the JSON, we respect the last name/key-chord set by the user. See [Modifying Actions](#modifying-actions)
+       in potential issues.
+4. Serialization
+    - `Action::ToJson()` and `ActionMap::ToJson()` should perform most of the work for us.
+    - See [Unbinding actions](#unbinding-actions) in potential issues.
+
+
+## UI/UX Design
+
+N/A
+
+## Capabilities
+
+N/A
+
+### Accessibility
+
+N/A
+
+### Security
+
+N/A
+
+### Reliability
+
+N/A
+
+### Compatibility
+
+N/A
+
+### Performance, Power, and Efficiency
+
+## Potential Issues
+
+### Copying the `ActionMap`
+
+The Settings UI needs `ActionMap::Copy()` so that it can bind XAML controls to the clone. However, the design
+above has two internal maps that own their `Action`s. Unfortunately, STL does not have multi-indexed data
+structures. To get around this issue, we can introduce internal action IDs to ensure we don't duplicate an action twice.
+
+```c++
+std::map<std::wstring, unsigned int> _NameMap;
+std::map<KeyChord, unsigned int> _KeyMap;
+
+// This could be stored as a `map<unsigned int, Action>`,
+//  but a vector provides this functionality much better.
+std::vector<Action> _ActionList;
+```
+
+With this design, the internal action ID can be saved as an `unsigned int`. This ID is never exposed.
+ When an `Action` is added to the `ActionMap`, we update `_ActionList`.
+
+Now that there are no duplicates between `_NameMap` and `_KeyMap`, creating a copy of `_NameMap` and
+ `_KeyMap` is trivial. We can iterate over `_ActionList` to create a copy of every `Action` stored
+ and save it to a new `_ActionList`.
+
+### Layering Actions
+
+We need a way to determine where an action came from to minimize how many actions we serialize when we
+ write to disk. This is a two part approach that happens as we're loading the settings
+ 1. Load defaults.json
+     - For each of the actions in the JSON...
+         - Construct the `Command` (basically the `Command::LayerJson` we have today)
+         - Add it to the `ActionMap`
+           - this should update `_NameMap`, `_KeyMap`, and `_ActionList` appropriately
+           - if the newly added name/key chord conflicts with a pre-existing one,
+              redirect `_NameMap` or `_KeyMap` to the newly added `Command` instead,
+              and update the conflicting one.
+2. Load settings.json
+     - Add a parent to `ActionMap`
+         - The purpose of a parent is to continue a search when the current `ActionMap`
+            can't find a `Command` for a query. The parent is intended to be immutable.

```

Comment:
Wait what now? This feels like a weird inverted way of phrasing this. We're taking the `ActionMap` from `defaults.json` and creating a _child_ `ActionMap`, then deserializing `settings.json` into that child, right?

Reply:
Correct, I'll clarify that in the spec rn.